README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->
[![Travis build status](https://travis-ci.org/venelin/PMMUsingSPLITT.svg?branch=master)](https://travis-ci.org/venelin/PMMUsingSPLITT) [![Coverage status](https://codecov.io/gh/venelin/PMMUsingSPLITT/branch/master/graph/badge.svg)](https://codecov.io/github/venelin/PMMUsingSPLITT?branch=master) [![CRAN\_Status\_Badge](http://www.r-pkg.org/badges/version/PMMUsingSPLITT?color=blue)](http://cran.r-project.org/web/packages/PMMUsingSPLITT) [![Downloads](http://cranlogs.r-pkg.org/badges/PMMUsingSPLITT?color=blue)](http://cran.rstudio.com/package=PMMUsingSPLITT)

PMMUsingSPLITT
==============

The goal of PMMUsingSPLITT is to provide a minimal example of how to use the SPLITT C++ library in an R-package. The package implements parallelized log-likelihood calculation of the univariate phylogenetic mixed model (PMM). The PMM is used in the comparative analysis of biological data originating from a set of living and/or extinct species to estimate the rate of phenotypic evolution resulting from genetic drift. The calculation of the log-likelihood of the model parameters, given a phylogenetic tree and trait data at the tips, is done using a quadratic polynomial representation of the log-likelihood function described in the article 'Parallel Likelihood Calculation for Phylogenetic Comparative Models: the 'SPLITT' C++ Library'. A preprint of the article is available from <https://doi.org/10.1101/235739>. The package provides an implementation in R (function 'AbcPMMLogLik') as well as a parallel implementation in C++ (function 'AbcPMMLogLikCpp') based on the 'SPLITT' library for serial and parallel lineage traversal of trees (<https://venelin.github.io/SPLITT/index.html>). The function 'MiniBenchmark' allows to compare the calculation times for different tree sizes. See [this guide](https://venelin.github.io/SPLITT/articles/SPLITTRcppModules.html) for a tutorial.

Installation
------------

You can install the released version of PMMUsingSPLITT from [CRAN](https://CRAN.R-project.org) with:

``` r
install.packages("PMMUsingSPLITT")
```

Examples
--------

-   Calculating the likelihood of the PMM for a given tree, data and model parameters.

``` r
library(ape)
library(PMMUsingSPLITT)
#> Loading required package: Rcpp

set.seed(10)

N <- 1000
x0 <- 0.1
alpha <- 1
theta <- 10
sigma2 <- 0.25
sigmae2 <- 1

tree <- rtree(N)

g <- rTraitCont(tree, model = "OU", root.value = x0,
                alpha = alpha, sigma = sqrt(sigma2),
                ancestor = FALSE)

x <- g + rnorm(n = N, mean = 0, sd = sqrt(sigmae2))

cat("logLikelihood using R:", AbcPMMLogLik(x, tree, x0, sigma2, sigmae2), "\n")
#> logLikelihood using R: -1551.016
cat("logLikelihood using R:", AbcPMMLogLikCpp(x, tree, x0, sigma2, sigmae2), "\n")
#> logLikelihood using R: -1551.016
```

-   Performing a benchmark to measure the likelihood calculation times using different parallelization strategies:

``` r
# N specifies the size of the phylogenetic tree. 
# Ntests specifies the number of executions in one time measurement 
# (the more the better, but also slower).
MiniBenchmark(N = 100, Ntests = 100)
#> Performing a mini-benchmark of the PMM log-likelihood calculation with 
#>       a tree of size N= 100 ;
#> Calling each likelihood calculation Ntests= 100  times ...
#> CPU:  Intel(R) Core(TM) i7-4850HQ CPU @ 2.30GHz 
#> OpenMP version:  201107 
#> Number of threads: 8 
#> Measuring calculation times...
#>    model                                            mode time.ms
#> 1    PMM                                      R (serial)    2.10
#> 2    PMM                                      C++ (AUTO)    0.04
#> 3    PMM              C++ (SINGLE_THREAD_LOOP_POSTORDER)    0.02
#> 4    PMM                 C++ (SINGLE_THREAD_LOOP_PRUNES)    0.02
#> 5    PMM                 C++ (SINGLE_THREAD_LOOP_VISITS)    0.02
#> 6    PMM                  C++ (MULTI_THREAD_LOOP_PRUNES)    0.23
#> 7    PMM                  C++ (MULTI_THREAD_LOOP_VISITS)    0.40
#> 8    PMM C++ (MULTI_THREAD_LOOP_VISITS_THEN_LOOP_PRUNES)    0.40
#> 9    PMM                  C++ (MULTI_THREAD_VISIT_QUEUE)    2.28
#> 10   PMM     C++ (MULTI_THREAD_LOOP_PRUNES_NO_EXCEPTION)    0.56
#> 11   PMM                        C++ (HYBRID_LOOP_PRUNES)    0.63
#> 12   PMM                        C++ (HYBRID_LOOP_VISITS)    0.39
#> 13   PMM       C++ (HYBRID_LOOP_VISITS_THEN_LOOP_PRUNES)    0.40
MiniBenchmark(N = 1000, Ntests = 10)
#> Performing a mini-benchmark of the PMM log-likelihood calculation with 
#>       a tree of size N= 1000 ;
#> Calling each likelihood calculation Ntests= 10  times ...
#> CPU:  Intel(R) Core(TM) i7-4850HQ CPU @ 2.30GHz 
#> OpenMP version:  201107 
#> Number of threads: 8 
#> Measuring calculation times...
#>    model                                            mode time.ms
#> 1    PMM                                      R (serial)    21.0
#> 2    PMM                                      C++ (AUTO)     0.1
#> 3    PMM              C++ (SINGLE_THREAD_LOOP_POSTORDER)     0.1
#> 4    PMM                 C++ (SINGLE_THREAD_LOOP_PRUNES)     0.1
#> 5    PMM                 C++ (SINGLE_THREAD_LOOP_VISITS)     0.2
#> 6    PMM                  C++ (MULTI_THREAD_LOOP_PRUNES)     0.9
#> 7    PMM                  C++ (MULTI_THREAD_LOOP_VISITS)     0.6
#> 8    PMM C++ (MULTI_THREAD_LOOP_VISITS_THEN_LOOP_PRUNES)     2.6
#> 9    PMM                  C++ (MULTI_THREAD_VISIT_QUEUE)    16.5
#> 10   PMM     C++ (MULTI_THREAD_LOOP_PRUNES_NO_EXCEPTION)     0.7
#> 11   PMM                        C++ (HYBRID_LOOP_PRUNES)     1.0
#> 12   PMM                        C++ (HYBRID_LOOP_VISITS)     0.1
#> 13   PMM       C++ (HYBRID_LOOP_VISITS_THEN_LOOP_PRUNES)     0.1
MiniBenchmark(N = 10000, Ntests = 10)
#> Performing a mini-benchmark of the PMM log-likelihood calculation with 
#>       a tree of size N= 10000 ;
#> Calling each likelihood calculation Ntests= 10  times ...
#> CPU:  Intel(R) Core(TM) i7-4850HQ CPU @ 2.30GHz 
#> OpenMP version:  201107 
#> Number of threads: 8 
#> Measuring calculation times...
#>    model                                            mode time.ms
#> 1    PMM                                      R (serial)   121.0
#> 2    PMM                                      C++ (AUTO)     0.6
#> 3    PMM              C++ (SINGLE_THREAD_LOOP_POSTORDER)     0.8
#> 4    PMM                 C++ (SINGLE_THREAD_LOOP_PRUNES)     0.6
#> 5    PMM                 C++ (SINGLE_THREAD_LOOP_VISITS)     0.8
#> 6    PMM                  C++ (MULTI_THREAD_LOOP_PRUNES)     0.3
#> 7    PMM                  C++ (MULTI_THREAD_LOOP_VISITS)     0.6
#> 8    PMM C++ (MULTI_THREAD_LOOP_VISITS_THEN_LOOP_PRUNES)     1.7
#> 9    PMM                  C++ (MULTI_THREAD_VISIT_QUEUE)   155.3
#> 10   PMM     C++ (MULTI_THREAD_LOOP_PRUNES_NO_EXCEPTION)     0.2
#> 11   PMM                        C++ (HYBRID_LOOP_PRUNES)     0.4
#> 12   PMM                        C++ (HYBRID_LOOP_VISITS)     1.5
#> 13   PMM       C++ (HYBRID_LOOP_VISITS_THEN_LOOP_PRUNES)     1.1
```


NEWS.md
# PMMUsingSPLITT 1.0.1

* Added a `NEWS.md` file to track changes to the package.


tests/test-all.R
library(testthat)

test_check("PMMUsingSPLITT")


tests/testthat/test-PMMLogLikelihoodSpeed.R
library(testthat)
context("Test MiniBenchmark(1000, 10) passes without errors")


library(PMMUsingSPLITT)

test_that("MiniBenchmark runs without errors", {
          expect_output(df <- PMMUsingSPLITT::MiniBenchmark(1000, 10), "Measuring calculation times...")
          expect_s3_class(df, "data.frame")
          })


tests/testthat/test-AbcPMMLogLikelihoodSpeed.R
library(testthat)
context("Test MiniBenchmark(1000, 10) passes without errors")


library(PMMUsingSPLITT)

test_that("MiniBenchmark runs without errors", {
          expect_output(df <- MiniBenchmark(1000, 10), "Measuring calculation times...")
          expect_s3_class(df, "data.frame")
          })


R/PMM.R
# PMM.R
# PMMUsingSPLITT
# 
# Copyright 2018 Venelin Mitov
# 
# This file is part of SPLITT: a generic C++ library for Serial and Parallel
# Lineage Traversal of Trees.
# 
# SPLITT is free software: you can redistribute it and/or modify
# it under the terms of the GNU Lesser General Public License as
# published by the Free Software Foundation, either version 3 of
# the License, or (at your option) any later version.
# 
# SPLITT is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Lesser General Public License for more details.
# 
# You should have received a copy of the GNU Lesser General Public
# License along with SPLITT.  If not, see
# <http://www.gnu.org/licenses/>.
# 
# @author Venelin Mitov


#' Calculate the PMM log-likelihood for a given tree, data and model parameters
#' @param x a numerical vector of size N, where N is the number of tips in tree
#' @param tree a phylo object
#' @param x0,sigma2,sigmae2 parameters of the PMM:
#' \describe{
#' \item{x0}{value at the root of tree excluding white noise;}
#' \item{sigma2}{unit-time variance increment of the heritable component;}
#' \item{sigmae2}{variance of the non-heritable component.}
#' }
#' @param ord an integer vector denoting the pruning order. This should be the 
#' result from calling `reorder(tree, order = "postorder", index.only = TRUE)`, 
#' which is also set as a default value. Can be passed as argument to speed-up 
#' the calculation.
#' @return the log-likelihood value.
PMMLogLik <- function(
  x, tree, x0, sigma2, sigmae2, 
  ord = reorder(tree, order = "postorder", index.only = TRUE)) {
  
  # number of tips in the tree
  N <- length(tree$tip.label)
  # total number of nodes in the tree (tips, internal nodes and root node)
  M <- nrow(tree$edge) + 1L
  # state variables for each node
  a <- b <- c <- rep(0.0, M)

  # indices of the rows in tree$edge in pruning order
  if(is.null(ord)) {
    ord <- reorder(tree, order = "postorder", index.only = TRUE)
  }

  for(o in ord) {
    # daughter node
    i <- tree$edge[o, 2]
    # parent node
    j <- tree$edge[o, 1]
    # branch length
    t <- tree$edge.length[o]

    if(i <= N) {
      # initialize a tip node
      a[i] <- -0.5 / sigmae2
      b[i] <- x[i] / sigmae2
      c[i] <- -0.5 * (x[i]*b[i] + log(2*pi*sigmae2))
    }

    d = 1 - 2*a[i]*sigma2*t

    # the order is important
    c[i] <- c[i] - 0.5*log(d) + 0.5*b[i]*b[i]*sigma2*t/d
    a[i] <- a[i] / d
    b[i] <- b[i] / d
    
    
    a[j] <- a[j] + a[i]
    b[j] <- b[j] + b[i]
    c[j] <- c[j] + c[i]
  }

  # for phylo objects, N+1 denotes the root node
  a[N+1]*x0^2 + b[N+1]*x0 + c[N+1]
}

#' Calculate the PMM log-likelihood for a given tree, data and model parameters using the Rcpp module
#' @inheritParams PMMLogLik
#' @param cppObject a previously created object returned by \code{\link{NewPMMCppObject}}
#' @param mode an integer denoting the mode for traversing the tree, i.e. serial vs parallel.
#' 
#' @return the log-likelihood value.
PMMLogLikCpp <- function(x, tree, x0, sigma2, sigmae2, 
                         cppObject = NewPMMCppObject(x, tree),
                         mode = getOption("SPLITT.postorder.mode", 0)) {
  abc <- cppObject$TraverseTree(c(sigma2, sigmae2), mode)
  abc[1]*x0^2 + abc[2]*x0 + abc[3]
}


#' Create an instance of the Rcpp module for a given tree and trait data
#'
#' @inheritParams PMMLogLik
#' @return an object to be passed as argument of the \link{PMMLogLikCpp} function.
#' @seealso \link{PMMLogLikCpp}
NewPMMCppObject <- function(x, tree) {
  PMMUsingSPLITT__TraversalTaskAbcPMM$new(tree, x[1:length(tree$tip.label)])
}


R/zzz.R
# zzz.R
# PMMUsingSPLITT
# 
# Copyright 2018 Venelin Mitov
# 
# This file is part of SPLITT: a generic C++ library for Serial and Parallel
# Lineage Traversal of Trees.
# 
# SPLITT is free software: you can redistribute it and/or modify
# it under the terms of the GNU Lesser General Public License as
# published by the Free Software Foundation, either version 3 of
# the License, or (at your option) any later version.
# 
# SPLITT is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Lesser General Public License for more details.
# 
# You should have received a copy of the GNU Lesser General Public
# License along with SPLITT.  If not, see
# <http://www.gnu.org/licenses/>.
# 
# @author Venelin Mitov


#' Rcpp module for the \code{TraversalTaskAbcPMM}-class
#' @name PMMUsingSPLITT__TraversalTaskAbcPMM
#' @aliases Rcpp_PMMUsingSPLITT__TraversalTaskAbcPMM-class
NULL

#' \code{TraversalAlgorithm}-type used in \code{AbcPMM}
#' @name PMMUsingSPLITT__AbcPMM__AlgorithmType
#' @aliases Rcpp_PMMUsingSPLITT__AbcPMM__AlgorithmType-class
NULL

#' Base class for \code{PMMUsingSPLITT::AbcPMM::AlgorithmType}
#' @name PMMUsingSPLITT__AbcPMM__TraversalAlgorithm
#' @aliases Rcpp_PMMUsingSPLITT__AbcPMM__TraversalAlgorithm-class 
#' @slot algorithm algorithm.
NULL

# loading the RCPP C++ modules
loadModule( "PMMUsingSPLITT__TraversalTaskAbcPMM", TRUE )


R/AbcPMM.R
# AbcPMM.R
# PMMUsingSPLITT
# 
# Copyright 2018 Venelin Mitov
# 
# This file is part of SPLITT: a generic C++ library for Serial and Parallel
# Lineage Traversal of Trees.
# 
# SPLITT is free software: you can redistribute it and/or modify
# it under the terms of the GNU Lesser General Public License as
# published by the Free Software Foundation, either version 3 of
# the License, or (at your option) any later version.
# 
# SPLITT is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Lesser General Public License for more details.
# 
# You should have received a copy of the GNU Lesser General Public
# License along with SPLITT.  If not, see
# <http://www.gnu.org/licenses/>.
# 
# @author Venelin Mitov


#' Calculate the PMM log-likelihood for a given tree, data and model parameters
#' @param x a numerical vector of size N, where N is the number of tips in tree
#' @param tree a phylo object
#' @param x0,sigma2,sigmae2 parameters of the PMM:
#' \describe{
#' \item{x0}{value at the root of tree excluding white noise;}
#' \item{sigma2}{unit-time variance increment of the heritable component;}
#' \item{sigmae2}{variance of the non-heritable component.}
#' }
#' @param ord an integer vector denoting the pruning order. This should be the 
#' result from calling `reorder(tree, order = "postorder", index.only = TRUE)`, 
#' which is also set as a default value. Can be passed as argument to speed-up 
#' the calculation.
#' @return the log-likelihood value.
AbcPMMLogLik <- function(
  x, tree, x0, sigma2, sigmae2, 
  ord = reorder(tree, order = "postorder", index.only = TRUE)) {
  
  # number of tips in the tree
  N <- length(tree$tip.label)
  # total number of nodes in the tree (tips, internal nodes and root node)
  M <- nrow(tree$edge) + 1L
  # state variables for each node
  a <- b <- c <- rep(0.0, M)

  # indices of the rows in tree$edge in pruning order
  if(is.null(ord)) {
    ord <- reorder(tree, order = "postorder", index.only = TRUE)
  }

  for(o in ord) {
    # daughter node
    i <- tree$edge[o, 2]
    # parent node
    j <- tree$edge[o, 1]
    # branch length
    t <- tree$edge.length[o]

    if(i <= N) {
      # initialize a tip node
      a[i] <- -0.5 / sigmae2
      b[i] <- x[i] / sigmae2
      c[i] <- -0.5 * (x[i]*b[i] + log(2*pi*sigmae2))
    }

    d = 1 - 2*a[i]*sigma2*t

    # the order is important
    c[i] <- c[i] - 0.5*log(d) + 0.5*b[i]*b[i]*sigma2*t/d
    a[i] <- a[i] / d
    b[i] <- b[i] / d
    
    
    a[j] <- a[j] + a[i]
    b[j] <- b[j] + b[i]
    c[j] <- c[j] + c[i]
  }

  # for phylo objects, N+1 denotes the root node
  a[N+1]*x0^2 + b[N+1]*x0 + c[N+1]
}

#' Calculate the AbcPMM log-likelihood for a given tree, data and model parameters using the Rcpp module
#' @inheritParams AbcPMMLogLik
#' @param cppObject a previously created object returned by \code{\link{NewAbcPMMCppObject}}
#' @param mode an integer denoting the mode for traversing the tree, i.e. serial vs parallel.
#' 
#' @return the log-likelihood value.
AbcPMMLogLikCpp <- function(x, tree, x0, sigma2, sigmae2, 
                         cppObject = NewAbcPMMCppObject(x, tree),
                         mode = getOption("SPLITT.postorder.mode", 0)) {
  abc <- cppObject$TraverseTree(c(sigma2, sigmae2), mode)
  abc[1]*x0^2 + abc[2]*x0 + abc[3]
}


#' Create an instance of the Rcpp module for a given tree and trait data
#'
#' @inheritParams AbcPMMLogLik
#' @return an object to be passed as argument of the \link{AbcPMMLogLikCpp} function.
#' @seealso \link{AbcPMMLogLikCpp}
NewAbcPMMCppObject <- function(x, tree) {
  PMMUsingSPLITT__TraversalTaskAbcPMM$new(tree, x[1:length(tree$tip.label)])
}


R/MiniBenchmark.R
# MiniBenchmark.R
# PMMUsingSPLITT
# 
# Copyright 2017 Venelin Mitov
# 
# This file is part of SPLITT: a generic C++ library for Serial and Parallel
# Lineage Traversal of Trees.
# 
# SPLITT is free software: you can redistribute it and/or modify
# it under the terms of the GNU Lesser General Public License as
# published by the Free Software Foundation, either version 3 of
# the License, or (at your option) any later version.
# 
# SPLITT is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Lesser General Public License for more details.
# 
# You should have received a copy of the GNU Lesser General Public
# License along with SPLITT.  If not, see
# <http://www.gnu.org/licenses/>.
# 
# @author Venelin Mitov

#' Perform a quick test of the PMM example likelihood calculation
#' @description This function runs a small benchmark to evaluate the SPLITT-package 
#' installation on a given computer. 
#' @param N number of tips in the test phylogenetic tree, default 10000
#' @param Ntests number of calculations within a call to sys.time (the resulting
#' times are averages Ntests calls). Default: 10.
#' @return a data.frame containing the response times for the PMM log-likelihood
#' calculation.
MiniBenchmark <- function(N = 10000, Ntests = 10) {
  
  # needed to pass the check
  speedupX <- NULL
  time.ms <- NULL
  
  cat("Performing a mini-benchmark of the PMM log-likelihood calculation with 
      a tree of size N=", 
      N, ";\nCalling each likelihood calculation Ntests=", 
      Ntests, " times ...\n")
  
  set.seed(10)
  
  x0 <- 0.1
  alpha <- 1
  theta <- 2.1
  sigma2 <- 0.25
  sigmae2 <- 1
  
  tree <- rtree(N)
  
  g <- rTraitCont(tree, model = "BM", root.value = x0, sigma = sqrt(sigma2),
                  ancestor = FALSE)
  
  x <- g + rnorm(n = N, mean = 0, sd = sqrt(sigmae2))
  
  if(R.version[['os']]=='linux-gnu') {
    # this only works on linux
    cpuInfo <- system("cat /proc/cpuinfo | grep 'model name' | uniq", intern = TRUE)
  } else if(Sys.info()["sysname"] == "Darwin") {
    # this only works on mac OS x
    cpuInfo <- system("sysctl -a -n machdep.cpu.brand_string", intern = TRUE)
  } else {
    cpuInfo <- "(I don't know how to figure it out on your OS, sorry.)"
  }
  
  cat("CPU: ", cpuInfo, "\n")
  
  ord <- reorder(tree, order = "postorder", index.only = TRUE)
  
  cppPMMObject <- NewAbcPMMCppObject(x, tree)
  
  if(cppPMMObject$algorithm$VersionOPENMP == 0) {
    cat("It seems that OpenMP was disabled during C++ compilation. For parallel
        tree traversal OpenMP should be available on your system and supported
        by your C++ compiler. Only serial tree traversal is possible with 
        the current SPLITT installation. Please, read the user guide for further 
        instructions on how to build the SPLITT library.")
  } else {
    cat("OpenMP version: ", cppPMMObject$algorithm$VersionOPENMP, "\n")    
  }
  
  cat("Number of threads:",  cppPMMObject$algorithm$NumOmpThreads, "\n")
  
  cat("Measuring calculation times...\n")
  
  # warm-up for mode AUTO
  for(t in 1:1000) AbcPMMLogLikCpp(x, tree, x0, sigma2, sigmae2, cppPMMObject, 0)
  
  tR <- system.time(for(t in seq_len(Ntests/10)) AbcPMMLogLik(x, tree, x0, sigma2, sigmae2, ord = ord))[3] / (Ntests/10)
  unname(tR)
  
  measureTimeAbcPMMCpp <- function(mode) {
    unname(
      system.time(
        for(t in seq_len(Ntests)) 
          AbcPMMLogLikCpp(x, tree, x0, sigma2, sigmae2, cppPMMObject, mode)
      )[3] / Ntests*1000
    )
  }
  
  resultsAbcPMM <- rbind(
    data.frame(model = "PMM", 
               mode = "R (serial)", 
               time.ms = tR*1000),
    data.frame(model = "PMM", 
               mode = "C++ (AUTO)", 
               time.ms = measureTimeAbcPMMCpp(0)),
    data.frame(model = "PMM", 
               mode = "C++ (SINGLE_THREAD_LOOP_POSTORDER)", 
               time.ms = measureTimeAbcPMMCpp(10)),
    data.frame(model = "PMM", 
               mode = "C++ (SINGLE_THREAD_LOOP_PRUNES)", 
               time.ms = measureTimeAbcPMMCpp(11)),
    data.frame(model = "PMM", 
               mode = "C++ (SINGLE_THREAD_LOOP_VISITS)", 
               time.ms = measureTimeAbcPMMCpp(12)),
    data.frame(model = "PMM", 
               mode = "C++ (MULTI_THREAD_LOOP_PRUNES)", 
               time.ms = measureTimeAbcPMMCpp(21)),
    data.frame(model = "PMM", 
               mode = "C++ (MULTI_THREAD_LOOP_VISITS)", 
               time.ms = measureTimeAbcPMMCpp(22)),
    data.frame(model = "PMM", 
               mode = "C++ (MULTI_THREAD_LOOP_VISITS_THEN_LOOP_PRUNES)", 
               time.ms = measureTimeAbcPMMCpp(23)),
    data.frame(model = "PMM", 
               mode = "C++ (MULTI_THREAD_VISIT_QUEUE)", 
               time.ms = measureTimeAbcPMMCpp(24)),
    data.frame(model = "PMM", 
               mode = "C++ (MULTI_THREAD_LOOP_PRUNES_NO_EXCEPTION)", 
               time.ms = measureTimeAbcPMMCpp(25)),
    data.frame(model = "PMM", 
               mode = "C++ (HYBRID_LOOP_PRUNES)", 
               time.ms = measureTimeAbcPMMCpp(31)),
    data.frame(model = "PMM", 
               mode = "C++ (HYBRID_LOOP_VISITS)", 
               time.ms = measureTimeAbcPMMCpp(32)),
    data.frame(model = "PMM", 
               mode = "C++ (HYBRID_LOOP_VISITS_THEN_LOOP_PRUNES)", 
               time.ms = measureTimeAbcPMMCpp(33))
  )
  
  rownames(resultsAbcPMM) <- NULL
  
  resultsAbcPMM
}
